CIP-0161? | Ouroboros Phalanx - Breaking Grinding Incentives

nhenin opened the pull request:
We propose a new Ouroboros extension titled **Ouroboros Φalanx** (pronounced *Phalanx*), which addresses two critical Cardano Problem Statements:

* [CPS-0021: Ouroboros Randomness Manipulation](https://github.com/cardano-foundation/CIPs/tree/master/CPS-0021)
* [CPS-0017: Settlement Speed](https://github.com/cardano-foundation/CIPs/tree/master/CPS-0017)

**Ouroboros Φalanx** enhances **Ouroboros Praos** by mitigating grinding attacks through a strategic increase in the **computational cost of leader election manipulation**. It extends the **nonce generation process from 1 to 2 epochs**, introducing a **verifiable delay function (VDF)** that remains efficient for honest participants, while making it **significantly more expensive for adversaries to bias the outcome**.

- Rendered [CPS-0021: Ouroboros Randomness Manipulation](https://github.com/nhenin/CIPs/tree/CIP-Ouroboros-Phalanx/CPS-0021) 
- Rendered [CPD-0021: Ouroboros Randomness Generation Sub-Protocol - The coin-flipping Problem](https://github.com/nhenin/CIPs/tree/CIP-Ouroboros-Phalanx/CPS-0021/CPD/README.md)
- Rendered [CIP-0161 Ouroboros Phalanx - Breaking Grinding Incentives](https://github.com/nhenin/CIPs/tree/CIP-Ouroboros-Phalanx/CIP-0161/README.md)

rphair reviewed (commented):
Thanks again @nhenin for another intriguing submission.  There are a number of editorial issues in common with your earlier & related submission:
* https://github.com/cardano-foundation/CIPs/pull/1009

so rather than a detailed elaboration of each point I'll just propose a discussion agenda of potential issues (with checklist, to update when they start to get resolved) & tag it for `Triage` at our next CIP meeting: https://hackmd.io/@cip-editors/117

As before & always, you will be welcome to attend to introduce the proposal: not for technical review, but certainly a good chance to discuss some preliminaries like writing style, title, target audience, scope & nomenclature.

---

* [ ] For CIP review & posting we will have to retitle this — for consistency with reader & developer expectations — to use the English spelling rather than the Greek / Cyrillic letter.

  * [ ] If the CIP title "Phalanx" appears differently in marketing materials then this stylisation can also be mentioned, if considered relevant by editors & reviewers (it might not be).
* [ ] Please consider the title change of the PR also for your document: or something else more concise and less figurative (e.g. CIP-0140 = `Ouroboros Peras - Faster Settlement`).
* [ ] `README` (in the main directory & every subdirectory) should be in all capitals.
* [ ] There is no `Security` CIP category, so this is just `Consensus`.
* [ ] It's OK to force push at first (i.e. while getting your branch in order) but please avoid it from then on, since during review it's important to individually discuss & attribute changes to the document; as requested here: https://github.com/cardano-foundation/CIPs/blob/master/CIP-0001/README.md#1a-authors-open-a-pull-request
* [ ] Illustrations are generally prohibited in an Abstract: often in academics & publishing web sites it's copied into other documents so an image creates problems for text copying & interpretation.
  * [ ] The term Phalanx is considered general knowledge and doesn't have to be defined, let alone illustrated.
* [ ] As with CPS-0021 I would editorially recommend (and insist unless other editors object) that the academically themed and styled content be moved a separate `README` in a subdirectory: with simply enough practical detail left behind in the _Specification_ so node implementors can understand what is involved in the implementation: without necessarily all the mathematical underpinnings.
  * [ ] Whatever remains should be formatted as Markdown headings _without_ the number.number.number format used in academic papers.  This is essential for:
    * consistency with other CIPs;
    * automatic provisioning of deep links into the document... which you have now, but should be indexed _by text_ rather than number (since sending links online is easier overall than referring by section number);
    * not having to renumber the whole document every time a section is inserted or deleted (maybe not a problem for you at this stage... but generally would be for maintainers & editors in the long term)
* [ ] The _Rationale_ also has numbered headings and these are problematic for the same reasons.
* [ ] The _Rationale_ is unlikely to be understood, as currently written, by anyone but a handful of industry experts... it would be very generous of you, and helpful to the community, to provide colloquially understandable justifications here & leave the brilliant but esoteric detail to a full "Rationale" in the "breakout" document already recommended.
* [ ] Editors have since recommended — and, I thought, you & I tentatively agreed on — adding a statement in or near the _Copyright_ section about how AI was used in the production of the document.  From the formatting, emphasis & embellishments it seems as certain as for CPS-0021 that AI has also produced the final product here.  As you yourself stressed the value of [before](https://github.com/cardano-foundation/CIPs/issues/1017#issuecomment-2807186382), it is now your responsibility to prove to your target audience that your ideas are humanly creative and original and weren't simply written by a machine.

---

I have already mentioned in CIP meetings — with consensus from Cardano standards veterans —that I am opposed to dumping academic papers into CIPs for all these reasons mentioned and more.  The beauty of the CIP platform is that it presents standards which _different elements of a community can understand and adopt_... and I am seeing none of that here yet.  

Whatever "equity" is perceived for Cardano agencies developing these research topics, and having a "CIP" posted (perhaps to attract funding, or simply to designate a hardfork), to fairly build that equity they will *also* have to keep in mind the community & target audience in writing & submitting these documents: not just splicing academic papers into the CIP template.

@nhenin I'll keep an eye out for revisions in the week ahead before the next CIP meeting — happily allowing one or more "force pushes" for ***substantial*** rearrangements & rewrites —but will wait to respond until my comments above can be validated by editors at or before that meeting.  I don't want these suggestions to be taken as personally as last time & so hope that pre-validating them with other editors & reviewers will help set a cooperative tone going forward & save time for everyone in the long run. 😎

rphair commented:
@Ryun1 @Crypto2099 @perturbing one thing at least one editor should set aside time for (not during the meeting!) is to read this material side-by-side with CPS-0021 to see if there are redundancies to suggest trimming (though I think @nhenin will already have thought of that) & generally to see how well the documents work together.

I would postpone this effort, though, until the "separate `README`" item on the checklist above has been fully considered & after any agreed-upon adjustments have been made.

polinavino commented on CIP-????/Readme.md:
```diff
@@ -0,0 +1,1815 @@
+
+##### 2.3.1. VDF Primitives
+
+To define Wesolowski VDF construction, we first introduce a serie of hash functions: $`\text{Hash}^{(n)}_\mathbb{N}`$, which samples random integers of $`n`$ bits, $`\text{Hash}^{(n)}_\text{prime}`$, which samples a random integer from the set of the first $`2^{2n}`$ prime numbers, and $`\text{Hash}_\mathbb{G}`$, which samples a random group element of the class group $`\mathbb{G}`$.
```
"introduce a series of hash functions"

polinavino commented on CIP-????/Readme.md:
```diff
@@ -0,0 +1,1815 @@
+| ------------------------- | ------------------------- |
+| **Input Parameters**      | <ul><li>$`\lambda \in \mathbb{N}`$ — Security parameter.</li><li>$\text{pre-}\eta_e \in \{0,1\}^{256}$ — 256-bit pre-nonce entropy for epoch $e$.</li></ul> |
+| **Steps**                 | <ol><li>Compute all VDF challenges:<br>$`\forall i\ x_i \leftarrow  \text{Hash}_\mathbb{G}(\text{pre-}\eta_e \|\|\ i) `$</li><li>Compute the initial aggregation nonce:<br>$`\alpha \leftarrow  \text{Hash}^{(\lambda)}_\mathbb{N}(x_1 \|\| \dots \|\| x_n) `$</li><li>Initializes the accumulators to the identity:<br>$`(\text{Acc}_x,\ \text{Acc}_y) \leftarrow (1_\mathbb{G}, ...
+| **Returned Output**       | $`(\text{Acc}_x,\ \text{Acc}_y,\ \alpha)`$ — Input and output accumulators and initial aggregation nocne.   |
```
nocne -> nonce

polinavino commented on CIP-????/Readme.md:
```diff
@@ -0,0 +1,1815 @@
+
+</center>
+
+Every time a VDF output is published on-chain, if no previous VDF output are missing, we shall update the accumulators' state using $`\texttt{VDF.Aggregation.Update}`$.
```
"if no previous VDF outputs are missing"

rrtoledo commented:
> Thanks again @nhenin for another intriguing submission. There are a number of editorial issues in common with your earlier & related submission:
> 
> * #1009
> 
> so rather than a detailed elaboration of each point I'll just propose a discussion agenda of potential issues (with checklist, to update when they start to get resolved) & tag it for `Triage` at our next CIP meeting: https://hackmd.io/@cip-editors/117
> 
> As before & always, you will be welcome to attend to introduce the proposal: not for technical review, but certainly a good chance to discuss some preliminaries like writing style, title, target audience, scope & nomenclature.
> 
> * [ ]  For CIP review & posting we will have to retitle this — for consistency with reader & developer expectations — to use the English spelling rather than the Greek / Cyrillic letter.
>   
>   * [ ]  If the CIP title "Phalanx" appears differently in marketing materials then this stylisation can also be mentioned, if considered relevant by editors & reviewers (it might not be).
> * [ ]  Please consider the title change of the PR also for your document: or something else more concise and less figurative (e.g. CIP-0140 = `Ouroboros Peras - Faster Settlement`).
> * [ ]  `README` (in the main directory & every subdirectory) should be in all capitals.
> * [ ]  There is no `Security` CIP category, so this is just `Consensus`.
> * [ ]  It's OK to force push at first (i.e. while getting your branch in order) but please avoid it from then on, since during review it's important to individually discuss & attribute changes to the document; as requested here: https://github.com/cardano-foundation/CIPs/blob/master/CIP-0001/README.md#1a-authors-open-a-pull-request
> * [ ]  Illustrations are generally prohibited in an Abstract: often in academics & publishing web sites it's copied into other documents so an image creates problems for text copying & interpretation.
>   
>   * [ ]  The term Phalanx is considered general knowledge and doesn't have to be defined, let alone illustrated.
> * [ ]  As with CPS-0021 I would editorially recommend (and insist unless other editors object) that the academically themed and styled content be moved a separate `README` in a subdirectory: with simply enough practical detail left behind in the _Specification_ so node implementors can understand what is involved in the implementation: without necessarily all the mathematical underpinnings.
>   
>   * [ ]  Whatever remains should be formatted as Markdown headings _without_ the number.number.number format used in academic papers.  This is essential for:
>     
>     * consistency with other CIPs;
>     * automatic provisioning of deep links into the document... which you have now, but should be indexed _by text_ rather than number (since sending links online is easier overall than referring by section number);
>     * not having to renumber the whole document every time a section is inserted or deleted (maybe not a problem for you at this stage... but generally would be for maintainers & editors in the long term)
> * [ ]  The _Rationale_ also has numbered headings and these are problematic for the same reasons.
> * [ ]  The _Rationale_ is unlikely to be understood, as currently written, by anyone but a handful of industry experts... it would be very generous of you, and helpful to the community, to provide colloquially understandable justifications here & leave the brilliant but esoteric detail to a full "Rationale" in the "breakout" document already recommended.
> * [ ]  Editors have since recommended — and, I thought, you & I tentatively agreed on — adding a statement in or near the _Copyright_ section about how AI was used in the production of the document.  From the formatting, emphasis & embellishments it seems as certain as for CPS-0021 that AI has also produced the final product here.  As you yourself stressed the value of [before](https://github.com/cardano-foundation/CIPs/issues/1017#issuecomment-2807186382), it is now your responsibility to prove to your target audience that your ideas are humanly creative and original and weren't simply written by a machine.
> 
> I have already mentioned in CIP meetings — with consensus from Cardano standards veterans —that I am opposed to dumping academic papers into CIPs for all these reasons mentioned and more. The beauty of the CIP platform is that it presents standards which _different elements of a community can understand and adopt_... and I am seeing none of that here yet.
> 
> Whatever "equity" is perceived for Cardano agencies developing these research topics, and having a "CIP" posted (perhaps to attract funding, or simply to designate a hardfork), to fairly build that equity they will _also_ have to keep in mind the community & target audience in writing & submitting these documents: not just splicing academic papers into the CIP template.
> 
> @nhenin I'll keep an eye out for revisions in the week ahead before the next CIP meeting — happily allowing one or more "force pushes" for _**substantial**_ rearrangements & rewrites —but will wait to respond until my comments above can be validated by editors at or before that meeting. I don't want these suggestions to be taken as personally as last time & so hope that pre-validating them with other editors & reviewers will help set a cooperative tone going forward & save time for everyone in the long run. 😎

Hi @rphair, thanks a lot for the review!

@nhenin is on PTO for the next two weeks, so I will try to answer when I can general comments on the documents. I was in charge of the cryptography, so I shall answer these directly however.

First of all, I am sorry for the acamedic style and paper dumping. I definitely should have reworked my report for it to be more accessible before merging it with the CIP. I will follow your advice and move most of the content in a new subdirectory, with academic references, while leaving the core points in the main document.

I will also update the document with the minor changes that you mentioned regarding the picture in the abstract, the hard links, numbered headings. As for the CIP's title, I will update it to the style you mentioned and update the PR name in consequence (I prefer the term "mitigating" rather than "avoiding").

I will force push any change after I get to go ahead from yourself and other editors.

rrtoledo commented on CIP-????/Readme.md:
```diff
@@ -0,0 +1,1815 @@
+
+In Wesolowski’s paper, the proof is defined as $g^{{2^T} /\ p}$ where $g$ is the challenge, $T$ the difficulty and $p$ is a prime number found by hashing the VDF input and output together.  
+The proof is thus a single group element that can be computed in at most $2\cdot T$ group operations and constant space, or $(1+1/s) \cdot T$ time where the number $s$ is both the number of processors and space while the verification takes $\text{log}_2 T$ scalar multiplications in $\mathbb{Z}/p$ and two small exponentiations in the group $\mathbb{G}$. The proving time can further be optimized to ...
+Wesolowski also presents aggregation and watermarking methods. The aggregation method does not consist in aggregating multiple proofs but computing a proof of several VDF challenges. This is done by batching all inputs and outputs together and creating a proof for this batched input. The watermarking is done by computing the VDF twice, once normally and another time on a combination of the challe ...
```
"The aggregation method consists in aggregating multiple pair of VDF challenges and outputs and generating a single proof showing correctness of each individual and aggregated pair."

Is this wording better?

nhenin commented:
Hi @rphair, thanks for your feedback. The title you suggested, *“Ouroboros Phalanx – Avoiding Grinding Attacks”*, doesn’t quite convey what Phalanx ultimately represents. I’ve gone ahead and shortened the title and removed the Greek letter, as you suggested.

As @rrtoledo mentioned, I’m currently on vacation and will address your points more thoroughly once I’m back (after August 17). I don’t have strong objections to most of your comments, except the following section:

> I have already mentioned in CIP meetings — with consensus from Cardano standards veterans —that I am opposed to dumping academic papers into CIPs for all these reasons mentioned and more. The beauty of the CIP platform is that it presents standards which different elements of a community can understand and adopt... and I am seeing none of that here yet.

> Whatever "equity" is perceived for Cardano agencies developing these research topics, and having a "CIP" posted (perhaps to attract funding, or simply to designate a hardfork), to fairly build that equity they will also have to keep in mind the community & target audience in writing & submitting these documents: not just splicing academic papers into the CIP template.

I’ll bring this up directly during your CIP weekly, but in the meantime, I want to crystallize a few thoughts here on this PR:

First, this CIP is *not* an academic paper. It’s a piece of engineering . So I’m genuinely surprised by how you're characterizing it, and I find myself wondering why. It’s a complete specification of an extension to Ouroboros (the Consensus category), and it clearly justifies how it addresses two specific CPS:

* [CPS-0017 / Settlement Speed](https://github.com/cardano-foundation/CIPs/tree/master/CPS-0017)
* [CPS-0021 / Ouroboros Randomness Manipulation](https://github.com/cardano-foundation/CIPs/tree/master/CPS-0021)

It follows both the CIP template and the guidelines you've defined.

1. So my first question is: are you questioning the **completeness** of the proposal? Are we saying there’s *too much* information for a change at the core of Cardano? The CIP template asks for a precise specification and a rationale that addresses the CPS, which is exactly what this document provides. If the issue is the document size, what’s “too big”? Splitting it into sub-Readmes, as you suggested, is a formatting comment. So: is your feedback mostly about formatting, or are there semantic concerns as well? That’s the first clarification I’d like to understand.
2. Regarding the *"community & target audience"* point — what do you see as the primary goal of a CIP? Is it to propose a technical solution to a problem, or to market that solution? I’m genuinely unsure based on your comment. This is, first and foremost, a technical document. The **Abstract** and **Motivation** sections provide the high-level framing, while the **Specification** includes visuals and diagrams specifically aimed at making the proposal more understandable  which, frankly, isn’t something you’d typically find in an academic paper. This document is designed to provide all the necessary information for another team (for example, the Intersect Consensus Team) to implement the proposal. The intent is to avoid a scenario where such a team would begin integration work without first having a solid foundation on whether the proposal truly addresses the issues it claims to solve. It’s written for engineers and consensus product owners first, but in a generous and accessible way so that any curious member of the community can follow and deepen their understanding of the topic. Again, I have the impression it aligns with the CIP guidelines, so I’m a bit puzzled by your comments here as well.

rphair commented:
@nhenin the point of these questions about target audience and "academic" vs. "community" appeal is mainly to answer this one question:

> Does the document have any relevance to those besides the author(s) and team members who produced it?

You have given us some reasons why it might, as well as a suggestion that the way it's organised already _is_ broken down by progressive levels of detail (though I personally don't see this yet).  Of course that's also up to fellow editors who are likely to be asking the same questions originally posted above.  Please don't "shoot the messenger" for my being the first one to bring these common review questions into the process.

For the sake of those interested engineers, scientists, developers, and community members who try to stay ahead of the 110+ CIPs and 15+ CPSs we have now (and growing) we try to make sure exhaustive detail is going to serve some practical purpose.  Editorially it _always_ helps for the deeper detail to be in sub-documents rather than in the main document, and it was very encouraging when @rrtoledo offered to help with that.

There is no "too much" information to be in a CIP: but its presentation is a courtesy to the reader and hence our (my) suggestion to sift it between general & specific into 2 separate documents.  I would happily wait for the results of the redraft suggested in https://github.com/cardano-foundation/CIPs/pull/1065#issuecomment-3143774005 because I think that same effort you performed for CPS-0021 produced something much more reader-friendly and accessible. 🙏

As with the related CPS, @nhenin if you pressed for merging this document as-is, you would likely get it.  As you say we do need to record these achievements of engineering and right now the CIP repository suits your subject matter & application best.  But without some editing you might then be in a situation of having produced a public document that would only be read by those who share & can follow your academic interest.  Since that's exactly what happens with "academic papers", _this_ is the basis for that comparison.

Since I don't want you to be "puzzled" anymore by suggestions like this, I would rather wait until Tuesday when we can get a broader range of eyes on the document.  If I have said anything unreasonable... or suggested refinements that aren't necessary or wouldn't benefit readers... that should come out fairly soon in the initial review.

rphair commented:
@nhenin @rrtoledo FYI tomorrow's regular CIP meeting (where this was on the agenda for `Triage`) has been postponed until the next biweekly CIP meeting slot on 19 August (https://hackmd.io/@cip-editors/117) due to general attendance at Rare Evo this week.  Editors were comparing some notes about this submission in messaging today & are planning a more well-rounded review here on GitHub in the days ahead.

nhenin commented:
Related to that point:

> \[ ] There is no Security CIP category, so this is just Consensus.

That was just a "reminder to myself", which I forgot to remove. I’ve now addressed it separately as an issue: [https://github.com/cardano-foundation/CIPs/issues/1076](https://github.com/cardano-foundation/CIPs/issues/1076)

rphair commented on CIP-????/Readme.md:
```diff
@@ -0,0 +1,1815 @@
+---
+CIP: ??
```
```suggestion
CIP: 161
```

rphair reviewed (commented):
@nhenin @rrtoledo thanks again for bringing this practical part of your achievement to the community & the CIP meeting yesterday.  The editors more generally confirmed the advantages of having this proposal exist at two levels:
* an entry level: allowing non-specialist readers to more easily confirm this proposal in governance decisions: regarding both inclusion in hard forks and potential funding propositions
* a full specification level: with everything as you have written to make this proposal understood & implemented by responsible stakeholders, as you originally conceived (with already given apologies for using the word "academic" to refer to the density of detail × esoteric subject matter)

It remains at your discretion whether this would have to exist as 2 separate documents, as opposed to a single document with an appendix.  As you know (cc @aldo555) we're still awaiting `cips.cardano.org` support for inter-document links to `README.md` files in subdirectories, as in CPS-0021 (though links to other-named files, like `SPEC.md` as @Crypto2099 suggested, should work at least in the top level directory):
* https://github.com/cardano-foundation/cf-cip-frontend/issues/51

So please @rrtoledo we will watch out for your proposed update as per your https://github.com/cardano-foundation/CIPs/pull/1065#issuecomment-3143774005 and whatever feedback you & @nhenin have taken on from yesterday's CIP meeting.  

Please let us know when it's been updated to your satisfaction and we will put it up for _Review_ on the agenda or tag `Last Check` depending on how happy everyone seems to feel with it at that point.  And thanks again to you & your team for the extraordinary work and special attention. 🙏

### in the short term

Please update the directory name to `CIP-0161` 🎉 and
* rename your top level file from `Readme.md` to `README.md`
* update the "Rendered" link in the OP to point to the new pathname
* (if & when creating another spec file as above) add another "rendered" link to this in your OP

It also came up at the meeting for CIP editors to preemptively address the potential complexity issues affecting longer & more detailed CIPs, which anyone following or advocating this proposal is welcome to help us address (in a planned update to CIP-0001) via this issue:
* https://github.com/cardano-foundation/CIPs/issues/1075

rphair reviewed (commented):
@nhenin I agree with practically everything that you've argued in https://github.com/cardano-foundation/CIPs/pull/1065#issuecomment-3214499130.  Time will tell about whether numbering the sections in this document works as well a couple years later as it does now... but in any case you're the author which makes you, by the default, the maintainer so it should be completely your decision what subsection titles to use.

Most importantly we need to correct the recently renamed directory from `CIP-161` to `CIP-0161` as requested in https://github.com/cardano-foundation/CIPs/pull/1065#pullrequestreview-3136122028.  This document is (candidate) CIP 161, also called `CIP-0161`, and the leading `0` is especially significant when you see the directory names arranged on this page: https://github.com/cardano-foundation/CIPs

I would like to give this one more comprehensive read before green-checking it — after some final adjustments as below — but would also like to see 1 or more of the other editors provide a detailed response to this material before that "final" review.  If too long goes by without that being done then please tag us individually here and/or on the Discord channel.

nhenin commented on CIP-161/README.md:
```diff
@@ -0,0 +1,1871 @@
+License: Apache-2.0
+---
+
+<details>
+<summary>🔍 Table of Contents</summary>
```
FYI, CIP-0140 (Peras) actually uses the same collapsed Table of Contents:
[https://github.com/cardano-foundation/CIPs/tree/master/CIP-0140](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0140)

That’s where I originally got the idea. You’re right that in CIP-0140 it appears between the Abstract and Motivation sections. I’ve also used a ToC in the CPS-0021 CPD [here](https://github.com/cardano-foundation/CIPs/blob/master/CPS-0021/CPD/README.md#table-of-contents)], though in that case it isn’t collapsed.

I believe that the ToC is somewhat redundant when you have the outline feature. But in the official CIP frontend, we don’t have that outline capability. That’s why the collapsible ToC still feels useful today, it fills a gap in the frontend’s navigation.

Do you know if adding an outline or ToC navigation feature is on the roadmap for the frontend? That would probably make the Manual Markdown ToC less necessary in the long run.

In the meantime, would it be acceptable if I place it the same way as in CIP-0140? I would say that ideally this should be handled as a frontend feature rather than manually, but until then I would propose we temporarily keep something consistent with previous CIPs (at least the consensus ones).

rphair replied:
wow @nhenin I didn't realise that CIP-0140 contained that divergence from our CIP-0001 dictated structure 😅 but apparently it passed review anyway and so could be classified as "no big deal" for others to implement a global ToC the same way.  Anything parsing Markdown will likely ignore HTML (which a collapsible section always will be), and if future parsers get into trouble on that feature then, as you say, it will be on the `Consensus` documents & we can work together to fix them across the board.

I was initially uncomfortable having another H2 (same level as Markdown `##`) for the ToC — since nonstandard major section titles generally disrupt CIP reading — but it does make sense here because otherwise the ToC ranks below the Abstract, which doesn't make sense conceptually.

Perhaps we will make this option available to other CIP/CPS writers and therefore offer it in CIP-0001.  @Ryun1 @Crypto2099 @perturbing should we create an issue for this?  In any case I'll bookmark it for my periodic CIP-0001 overhauls and will leave this unresolved in case you guys want to add anything.

> Do you know if adding an outline or ToC navigation feature is on the roadmap for the frontend? That would probably make the Manual Markdown ToC less necessary in the long run.

The CF developer @aldo555 who currently maintains [cardano-foundation/cf-cip-frontend](https://github.com/cardano-foundation/cf-cip-frontend) recently asked the CIP editors if there were any features that were needed.  I'll let him know immediately that we need Markdown section outlining: as we have on GitHub and we used to have on the Developer Portal when CIPs where hosted there until a year or so ago.  If & when they do that, I'll leave it to you @nhenin (or maybe @bwbush) to submit a PR to remove those collapsible sections.

nhenin replied:
@rphair , I have added this issue to the list as well FYI : https://github.com/cardano-foundation/cf-cip-frontend/issues/53

rphair commented on CIP-0161/README.md:
```diff
@@ -0,0 +1,1871 @@
+Ouroboros Phalanx therefore represents a **complementary advancement**: reinforcing Cardano’s consensus security while improving performance, and ensuring the network remains robust against future adversarial strategies.
+
+<details>
+<summary><h2>🔍 Table of Contents</h2></summary>
```
```suggestion
<summary><h2>Table of Contents</h2></summary>
```
p.s. to https://github.com/cardano-foundation/CIPs/pull/1065#discussion_r2297761342 - I do believe, if only for consistency with CIP-0140 as you mention @nhenin, that the icon should be removed: since we are ushering this in with the major headings which don't have emojis, and also because the blank unexpanded _Table of Contents_ already invites clicking on it with out the call-to-action emoji.

nhenin commented:
@Ryun1 @Crypto2099 @perturbing – will you be able to review this PR soon? We’re holding on the merge to proceed with downstream communications and the next set of actions.

Ryun1 commented on CPS-0017/README.md:
```diff
@@ -7,6 +7,10 @@ Authors:
   - Hans Lahe <hans.lahe@iohk.io>
+Proposed Solutions: 
+	["CIP-0140 : Ouroboros Peras - Faster Settlement"
+	,"CIP-???? : Ouroboros Φalanx: Breaking the Economics of Grinding Attacks"
```
```suggestion
	,"CIP-161 : Ouroboros Phalanx : Breaking the Economics of Grinding Attacks"
```

Ryun1 commented on CIP-0161/README.md:
```diff
@@ -0,0 +1,1871 @@
+- When enough committee members participate, it achieves **rapid settlement**.  
+- When they do not (e.g., if too many operators are offline), Peras **gracefully falls back to Praos**.  
+
+In these fallback situations, the network still relies on Praos’ guarantees—precisely where Phalanx becomes relevant as a **complementary defense layer**. Phalanx ensures that even when Peras cannot certify blocks, the protocol still benefits from:  
```
im curious
do we know the likelihood of these fallback situations?

nhenin replied:
Peras would only fall back to Praos if there were massive network disruption or an attack by a 25% adversary.

nhenin replied:
I will add it

rphair commented on CPS-0021/README.md:
```diff
@@ -6,8 +6,8 @@ Status: Open
-    - Peter Gaži <peter.gazi@iohk.io>
-Proposed Solutions: []
+Proposed Solutions: 
+  ["CIP-???? : Ouroboros Φalanx: Breaking the Economics of Grinding Attacks"]
```
yes @nhenin, it's important to keep the titles consistent (i.e. without the Greek letter) & also spell the CIP tags consistently (cc @Ryun1) and to use the same labelling syntax already in use for other CIPs (see [header preamble](https://github.com/cardano-foundation/CIPs/blob/master/CIP-0001/README.md#header-preamble)):
```suggestion
    - "CIP-0161: Ouroboros Phalanx": https://github.com/cardano-foundation/CIPs/pull/1065
```
The creation of a set by using `[]` around a nonempty list would be valid YAML but contrary to the syntax all other CIPs are using at this point.

rphair commented:
yes, it's fine @nhenin since you force-pushed a branch that contained all commits that we already had in the commit history.  In the interest of brevity we often have to say "no force push" even though what we really mean is "don't force push a brand new branch that has your updated content without all the commits that created it" 🤓
- @Ryun1 @Crypto2099 @perturbing maybe we should update the CIP-0001 language accordingly?  (I'll put that review on my TODO)

We don't have a "deadline for merging" at any time, and since our CIP meeting is biweekly the `Last Check` status would next be applied at this meeting next week: https://hackmd.io/@cip-editors/119 — so at this point we are just waiting for @perturbing @Crypto2099 so have a comprehensive look & review.  If they have no objections by next Tuesday we can merge it at that meeting... or earlier if they indicate approval or consent here.

perturbing commented:
I apologize for the late review; I have it on the to-do list for later today

perturbing commented on CIP-0161/README.md:
```diff
@@ -0,0 +1,1871 @@
+
+![alt text](./image/Praos-vs-Phalanx-Highl-Level.png)
+
+1. The **stake distribution stabilization phase** is shifted **back by one epoch :** The **active** **stake distribution** $`\mathbf{SD}_e`$ used for leader election is now derived from the **end of $epoch_\text{e-3}$** instead of **$epoch_\text{e-2}$**  as in the original Praos protocol.  
```
@bwbush, just to be certain, how would such a change impact Leios?

Given that both Leios and Phalanx have been deemed secure additions to Proas, is their combination also considered?

rrtoledo commented on CIP-0161/README.md:
```diff
@@ -0,0 +1,1871 @@
+
+## Motivation: why is this CIP necessary?
+
+This proposal strengthens Cardano’s consensus protocol (Ouroboros Praos) against a class of weaknesses known as *grinding attacks*. These attacks allow adversaries to bias the randomness used in block leader elections, statistically slowing settlement and weakening security guarantees.
```
It does not weaken security guarantees, it is the opposite of that: it ensures security even when weaker assumptions.

rrtoledo commented on CIP-0161/README.md:
```diff
@@ -0,0 +1,1871 @@
+
+The improvement introduces an additional computation step that is lightweight for honest participants but significantly more expensive for attackers, making grinding attacks economically infeasible.
+
+### Recommended Configuration
```
I suppose you are asking about "Recommending Confguration"?

The design is indeed flexible, there is a spectrum of parameters that we can choose from as well as some optional features.

Here we wrote down some configurations to give to the reader an idea of what the protocol could look like to facilitate discussions and motivate why this CIP is necessary but also reasonnable.

nhenin replied:
@perturbing We have deliberately included in this section all the information a non-expert needs to understand this proposal (right before the warning you liked 🙂). This was added in response to the feedback we received at the start of this [PR](https://github.com/cardano-foundation/CIPs/pull/1065#issuecomment-3214499130).

rphair replied:
My apologies @nhenin — I think I'd inadvertently been looking a different part of the document when I made https://github.com/cardano-foundation/CIPs/pull/1065#discussion_r2352591456.  In any case everything runs smoothly after your last https://github.com/cardano-foundation/CIPs/pull/1065/commits/c756c345acb1e49b51dce01073d5e4956fb16de1 so this is definitely resolved (cc @perturbing).

rrtoledo commented on CIP-0161/README.md:
```diff
@@ -0,0 +1,1871 @@
+
+## Specification
+
+The core principle of the proposed protocol change is to **substantially escalate the computational cost of each grinding attempt for an adversary**. To achieve this, every honest participant is required to perform a designated computation for each block they produce over an epoch (**432,000 slots - 5 days**). Consequently, an adversary attempting a grinding attack must **recompute these operatio ...
```
I cannot say about DeltaQ, but about Praos I think this change may be (perhaps not fully) orthogonal.

As we are asking only the block producers to perform some extra computations, these changes should not affect the input or endorsement blocks, only the ranking ones - leaving the throughput unchanged. 

The latency could have worsen but, as you noted, we made it so that the computational overhead can be done in advance, so with a good scheduling there should not be a degradation in performances. 

Sure we can highlight that the computation overhead can be precomputed. I'll add a suggestion in this thread.

rrtoledo replied:
```suggestion
The core principle of the proposed protocol change is to **substantially escalate the computational cost of each grinding attempt for an adversary**. To achieve this, every honest participant is required to perform a designated computation for each block they produce over an epoch (**432,000 slots - 5 days**) - note that this computation can be preprocessed locally at the beginning of the epoch. Consequently, an adversary attempting a grinding attack must **recompute these operations for every single attempt**, while being **constrained by the grinding window**, which dramatically increases the resource expenditure. By enforcing this computational burden, we **drastically reduce the feasible number of grinding attempts** an adversary with a fixed resource budget can execute, making randomness manipulation **more expensive and significantly less practical**.
```

rrtoledo commented on CIP-0161/README.md:
```diff
@@ -0,0 +1,1871 @@
+
+![alt text](./image/Praos-vs-Phalanx-Highl-Level.png)
+
+1. The **stake distribution stabilization phase** is shifted **back by one epoch :** The **active** **stake distribution** $`\mathbf{SD}_e`$ used for leader election is now derived from the **end of $epoch_\text{e-3}$** instead of **$epoch_\text{e-2}$**  as in the original Praos protocol.  
```
We did not write about the transition indeed.

One way to do it would be to have a hybrid period where the seed at epoch e we compute both the randomness of epoch e+1 normally as well as the randomness of epoch e+2 with Phalanx.

Another way would be to not mix the two phases and directly use Phalanx, in which case the randomness of epoch e+1 would be computed as the fallback of Phalanx (since no computation would have been submitted).

rrtoledo replied:
If you are happy with adding this information, I'll add a subsection with this in the specifications. In term of complexity, the second option seem simpler but I have to double check the security of it.

Is there anything else you think is missing for the transition?

rphair commented:
@nhenin we follow a convention of applying a `Waiting for Author` tag when we can document that the author needs more time to respond to reviews **and/or** to develop a proposal before it can be reviewed further (e.g. [Merkle Root](https://github.com/cardano-foundation/CIPs/pull/964) [cc @aslesarenko @colll78]) and [Light Client Protocols](https://github.com/cardano-foundation/CIPs/pull/942) [cc @cleanerm5])).
* This categorisation is described here, which is the closest thing to an SLA we currently have about it: https://github.com/cardano-foundation/CIPs/wiki/301.-State-tagging#state-tag-details

Conversely: we don't apply that tag when the CIP PR is in a state when general review is still possible... and therefore potentially applicable to whatever deliberations the author may currently be going through before progressing their proposal (e.g. [Nested Transactions](https://github.com/cardano-foundation/CIPs/pull/862) [cc @lehins] who's taken it over from @polinavino).

Separately: when subsequent CIPs go beyond, or redefine, the scope of an apparently related CPS (e.g. [Intents for Cardano](https://github.com/cardano-foundation/CIPs/pull/779) [cc @michaelpj]) the expectation — from authors, not editors — is that authors or advocates of those CIPs will either (with the applicable CPS remaining closed or tagged as stalled / deprecated until then):
* recognise the connection with some already submitted CPS, open that PR (if closed) & offer to maintain it (I did exactly this with https://github.com/cardano-foundation/CIPs/pull/947);
* submit a new CPS that redefines the problem in a more readily & currently applicable way.

None of these conditions apply to Phalanx.  It's only stalled because _both_ these conditions are true:
1. half the CIP editors haven't had the time to review it to their satisfaction;
2. the same community of `Consensus` reviewers, and those in related disciplines, have not endorsed this with the same level of attention that's been paid to Leios... keeping in mind it's still only 2 days since I posted [this wider call for review](https://github.com/cardano-foundation/CIPs/pull/1065#issuecomment-3305740016) and even Leios took a couple weeks to pull the review train up to full speed after the author's similar request (https://github.com/cardano-foundation/CIPs/pull/1078#issuecomment-3249162612: cc @will-break-it).

Therefore @nhenin if you consider # 1 to be a bottleneck, since technically this proposal is ready to merge with its current 2 editor endorsements, the best way to move it forward — if editors can observe evidence that subject matter experts are generally providing favourable peer review — would be to use your own influence to make progress on # 2.

rphair commented:
@nhenin cross-referencing this Cardano Forum posting which I believe could be effective as another call for feedback & review: https://forum.cardano.org/t/join-us-for-a-live-discussion-on-cip-0161-ouroboros-phalanx/149902
